# Test Assignment. Junior Data Analyst 
## Damir Umetbaev

Для того чтобы не обсчитывать ресурсоемкий процесс подготовки данных каждый раз при перезапуске кода, вынес его в отдельный файл, и результат сохранил в csv

In [3]:
import pandas as pd
import numpy as np
import json

In [4]:
df = pd.read_csv('us events ad_revenue filtered 03.02-07.02.csv')

## Check for Duplicates

In [5]:
df_check = df.drop_duplicates(subset=['device_locale'])
df_check.shape

(9, 16)

In [6]:
df_check = df.drop_duplicates(subset=['event_name'])
df_check.shape

(1, 16)

В случае когда строка в полученной таблице всего одна, значение поля (к примеру event_name) одинакого для всех строк. Это значит что поле информации не несет. 
Такие поля данных (столбцы) я удалю из датасета, с целью оптимизации производительности ввиду большого размера датасета.
Строки которые будут удалены: event_name, country_iso_code


## Check for missing values

In [23]:
df.isnull().sum()

profile_id                     14
os_version                      0
device_manufacturer             0
device_model                    0
device_type                     0
device_locale                   0
event_name                      0
event_json                      0
event_datetime                  0
event_timestamp                 0
event_receive_datetime          0
event_receive_timestamp         0
connection_type                 0
operator_name               18879
country_iso_code                0
city                       323708
dtype: int64

Информация о городе отсутствует в 323708 случаях из 860079

## Parse json data into columns

In [10]:
dictionary = df.event_json.apply(json.loads)
dictionary_df = pd.json_normalize(dictionary)
dictionary_df.isnull().sum()

Информация о доходе присутствует всегда ->
Процент доходов содержащий информацию о городе = 62% [(860079-323708)/860079]

In [13]:
df_parsed = pd.concat([df, dictionary_df], axis=1)
# delete not relevant columns
df_parsed = df_parsed.drop(['event_json','event_name', 'country_iso_code'], axis=1)
df_parsed.to_csv('df_parsed.csv')

profile_id os_version device_manufacturer         device_model  \
0       JLjpGzDHBQm         11             Samsung        Galaxy Tab S6   
1       JLjpGzDHBQm         11             Samsung        Galaxy Tab S6   
2       JLkVacrzHY6         11              Google              Pixel 2   
3       JMC6wG3cX0f         10             Samsung            Galaxy S9   
4       JLzWWTJfi1b         10             Samsung            Galaxy S9   
...             ...        ...                 ...                  ...   
860074  JLSQY2nx1NQ         12             Samsung  Galaxy S21 Ultra 5G   
860075  JLKWwju79CA         12             Samsung             SM-G991U   
860076  JLKWwju79CA         12             Samsung             SM-G991U   
860077  JLKWwju79CA         12             Samsung             SM-G991U   
860078  JLKWwju79CA         12             Samsung             SM-G991U   

       device_type device_locale       event_datetime  event_timestamp  \
0           tablet         en_US  2022-02-04 20:02:28       1643994148   
1           tablet         en_US  2022-02-04 20:02:45       1643994165   
2            phone         en_US  2022-02-05 17:53:05       1644072785   
3            phone         en_US  2022-02-06 17:09:01       1644156541   
4            phone         en_US  2022-02-06 17:15:38       1644156938   
...            ...           ...                  ...              ...   
860074       phone         en_US  2022-02-03 08:29:27       1643866167   
860075       phone         en_US  2022-02-03 02:12:51       1643843571   
860076       phone         en_US  2022-02-03 02:13:06       1643843586   
860077       phone         en_US  2022-02-03 02:13:17       1643843597   
860078       phone         en_US  2022-02-03 03:11:44       1643847104   

       event_receive_datetime  event_receive_timestamp  ...   ads_ltv  \
0         2022-02-06 17:58:19               1644159499  ...  0.487877   
1         2022-02-06 17:58:19               1644159499  ...  0.488387   
2         2022-02-06 16:59:29               1644155969  ...  0.678479   
3         2022-02-06 17:09:24               1644156564  ...  0.013894   
4         2022-02-06 17:16:16               1644156976  ...  0.060571   
...                       ...                      ...  ...       ...   
860074    2022-02-03 08:29:37               1643866177  ...  1.007506   
860075    2022-02-03 02:14:48               1643843688  ...  0.277541   
860076    2022-02-03 02:14:48               1643843688  ...  0.280321   
860077    2022-02-03 02:14:48               1643843688  ...  0.283101   
860078    2022-02-03 21:12:34               1643911954  ...  0.285881   

         ecpm_sma normalized_ecpm_sma     value currency  ad_country_code  \
0        0.511508            0.511508  0.000510      USD               US   
1        0.511389            0.511389  0.000510      USD               US   
2        0.928597            0.928597  0.002386      USD               US   
3       13.782633           13.893784  0.013894      USD               US   
4        1.660282            1.660282  0.006828      USD               US   
...           ...                 ...       ...      ...              ...   
860074   0.080914            0.080914  0.000020      USD               US   
860075   3.379858            3.379858  0.006740      USD               US   
860076   3.319878            3.319878  0.002780      USD               US   
860077   3.265896            3.265896  0.002780      USD               US   
860078   3.217312            3.217312  0.002780      USD               US   

             network_name        ad_unit_id  \
0            Google AdMob  64c6b1114ea8c1cc   
1            Google AdMob  64c6b1114ea8c1cc   
2       APPLOVIN_EXCHANGE  64c6b1114ea8c1cc   
3                AppLovin  3ab7705b07ebb75f   
4            Google AdMob  64c6b1114ea8c1cc   
...                   ...               ...   
860074  APPLOVIN_EXCHANGE  64c6b1114ea8c1cc   
860075       Google AdMob  64c6b1114ea8c1